### PR TITLE
DM-11952: Proposed change-controlled document workflow

### PR DESCRIPTION
**Pull request by jonathansick**
https://developer.lsst.io/v/DM-11952/docs/change-controlled-docs.html

### CODE REVIEW

File: docs/change-controlled-docs.rst [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+.. _ccd-submit:
+
+Submitting a document to the CCB
+================================
+
+Establishing a new baselined version of the document begins with submitting it to the CCB.
+The release manager is responsible for orchestrating this process on behalf of DM.
+
+1. Choose a commit on ``master`` where the document's content is ready for CCB review.
+2. Create a DocuShare tag at the commit.
```
Comment (timj) [inline docs/change-controlled-docs.rst]: No. This has to be step 3. Docushare has its own version numbers for uploading (not to be confused with Version handles in Docushare which are global document versions) and you should not assume a number but should upload the draft to docushare and then look at what you got (it should be the next one but why risk it?).

File: docs/change-controlled-docs.rst [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+   DocuShare tags are formatted as ``docushare-vN``, where ``N`` is the next-available document version number in DocuShare.
+3. Upload the PDF built by CI for the ``docushare-vN`` tag to DocuShare.
+   By definition, this uploaded document should become version ``N`` in DocuShare.
+4. For LDM documents, create a DM RFC.
+   For project documents (LSE), create an LCR.
```
Comment (timj) [inline docs/change-controlled-docs.rst]: LSE or LPM

File: docs/change-controlled-docs.rst [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+================================
+
+Establishing a new baselined version of the document begins with submitting it to the CCB.
+The release manager is responsible for orchestrating this process on behalf of DM.
+
+1. Choose a commit on ``master`` where the document's content is ready for CCB review.
+2. Create a DocuShare tag at the commit.
+   DocuShare tags are formatted as ``docushare-vN``, where ``N`` is the next-available document version number in DocuShare.
+3. Upload the PDF built by CI for the ``docushare-vN`` tag to DocuShare.
+   By definition, this uploaded document should become version ``N`` in DocuShare.
```
Comment (timj) [inline docs/change-controlled-docs.rst]: No need for this sentence if we've flipped the order of operation.

File: docs/change-controlled-docs.rst [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+   Example of Git branches during document development.
+   Staff write and edit the document on individual ticket branches that they merge to master, following DM's regular code workflow.
+
+.. _ccd-submit:
+
+Submitting a document to the CCB
+================================
+
+Establishing a new baselined version of the document begins with submitting it to the CCB.
+The release manager is responsible for orchestrating this process on behalf of DM.
```
Comment (timj) [inline docs/change-controlled-docs.rst]: "on behalf of DM" suggests this sentence refers to LSST CCB. Should probably capitalize Release Manager. What does "orchestrating" mean? Which steps below are mediated by the Release Manager? Not everyone who writes an LDM will be able to upload to docushare. They may have to ask someone from DM Systems Engineering to help (or the Release Manager).

File: docs/change-controlled-docs.rst [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+
+1. Choose a commit on ``master`` where the document's content is ready for CCB review.
+2. Create a DocuShare tag at the commit.
+   DocuShare tags are formatted as ``docushare-vN``, where ``N`` is the next-available document version number in DocuShare.
+3. Upload the PDF built by CI for the ``docushare-vN`` tag to DocuShare.
+   By definition, this uploaded document should become version ``N`` in DocuShare.
+4. For LDM documents, create a DM RFC.
```
Comment (timj) [inline docs/change-controlled-docs.rst]: Do we still have the RFC process in the dev guide to link to? I think these should be immediately set to FLAGGED state.

File: docs/change-controlled-docs.rst [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+   This release branch initially has no new commits on it.
+
+.. figure:: ccd-submit.svg
+   :name: fig-ccd-submit
+   :align: center
+   :target: ccd-submit.svg
+   :alt: Git commit tree showing a docushare-v1 tag added to the head of the master branch.
+
+   Example Git tree showing document submission to the CCB.
+   The release manager creates a ``docushare-v1`` tag from the head of the ``master`` branch, since ``1`` is the next-available version number for the document in DocuShare.
```
Comment (timj) [inline docs/change-controlled-docs.rst]: It is highly unlikely that 1 will ever be used because when you first create a document in docushare for an LDM you don't usually have a document to upload to you ask docushare to create a stub. That will be v1.

Comment (timj) [inline docs/change-controlled-docs.rst]: Are you saying that by "orchestrate" you mean the Release Manager is doing all these steps?

File: docs/change-controlled-docs.rst [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+   In this case, the submission is ``RFC-1`` so the release manager creates the ``tickets/RFC-1`` release branch from ``docushare-v1`` tag.
+
+.. _ccd-edit:
+
+Addressing CCB feedback
+=======================
+
+The CCB may request amendments before the document can be baselined.
+
+1. Create a ticket branch from the ``tickets/RFC-N`` or ``tickets/LCR-N`` release branch.
```
Comment (timj) [inline docs/change-controlled-docs.rst]: we create a JIRA story even if the feedback is to do a one line change? Or we allow commits to the RFC/LCR branch directly?

File: docs/change-controlled-docs.rst [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+2. Commit amendments to that ticket branch and peer review.
```
Comment (timj) [inline docs/change-controlled-docs.rst]: Do we need peer review if the CCB have requested a change? The CCB will be about to review the change anyhow. When dealing with major feedback that was not clear peer review would be useful but are we mandating it?

File: docs/change-controlled-docs.rst [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+3. Rebase and merge the ticket branch onto the ``tickets/RFC-N`` or ``tickets/LCR-N`` release branch.
+4. Multiple ticket branches can be created and merged to the release branch if the CCB's requests are being addressed by multiple people working on separate parts of the document.
+5. The release manager creates a new DocuShare tag (``docushare-vN``) from the head of the release branch, uploads the revised document to DocuShare, and notifies the CCB.
```
Comment (timj) [inline docs/change-controlled-docs.rst]: fix ordering of steps.

File: docs/change-controlled-docs.rst [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+
+.. figure:: ccd-amend.svg
+   :name: fig-amend
+   :align: center
+   :target: ccd-amend.svg
+   :alt: Git commit tree showing amendments and resubmission to a CCB.
+
+   Example Git tree showing how CCB feedback is addressed on a release branch.
+   Two individuals address CCB feedback by creating separate ticket branches (``tickets/DM-3`` and ``tickets/DM-4``) from the release branch (``RFC-1``) that are merged back to the ``RFC-1`` release branch
+   The release manager resubmits the amended document to the CCB by creating a docushare tag (``docushare-v2``) and uploading version 2 of the document to DocuShare.
```
Comment (timj) [inline docs/change-controlled-docs.rst]: For DM CCB we *could* say that the review process can use the .lsst.io RFC ticket branch version. Uploading the first draft to docushare and the final accepted version to docushare seems enough to me to satisfy LPM-51. I don't think LPM-51 requires that intermediate versions under review go through docushare. LSST CCB may not be as comfortable with that approach.

If the release manager is mediating all this by definition I don't want to add extra work that's not needed. If the author of the LDM thinks they have dealt with the review comments then they should be able to say so on the RFC and point at lsst.io without involving the release manager.

File: docs/change-controlled-docs.rst [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+
+.. _ccd-release:
+
+Releasing a new baselined version of the document
+=================================================
+
+When the CCB baselines (accepts) the new document, the project librarian designates the baselined document with a semantic version.
```
Comment (timj) [inline docs/change-controlled-docs.rst]: The librarian is only involved for LSST CCB releases. The release manager can determine the semantic version for DM CCB.

File: docs/change-controlled-docs.rst [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+The DM release manager is responsible for releasing a document.
```
Comment (timj) [inline docs/change-controlled-docs.rst]: Although I'd like the DM Systems Engineer to also be allowed to release my own documents...

File: docs/change-controlled-docs.rst [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+
+1. From the head of the release branch (``tickets/RFC-N`` or ``tickets/LCR-N``), add a commit that:
+
+	- Updates the semantic version of the document to the one designated by the project librarian.
+	- Updates the change history table to reflect the new version.
+	- Fixes the document's date to the date of baselining.
+	- Removes the "draft" watermark (for example, by removing ``draft`` from the document class's arguments).
+
+2. Create a new DocuShare tag (``docushare-vN``) and upload the document to DocuShare.
```
Comment (timj) [inline docs/change-controlled-docs.rst]: Nope 😄

File: docs/change-controlled-docs.rst [hunk 14]
```diff
@@ -0,0 +1,233 @@
...
+
+3. Once the project librarian has set that DocuShare version as the preferred version, create a semantic version tag (formatted ``v<major>.<minor>``) at the same commit as the corresponding DocuShare tag.
```
Comment (timj) [inline docs/change-controlled-docs.rst]: librarian only for LSST CCB.

File: docs/change-controlled-docs.rst [hunk 15]
```diff
@@ -0,0 +1,233 @@
...
+.. figure:: ccd-release.svg
+   :name: fig-release
+   :align: center
+   :target: ccd-release.svg
+   :alt: Git commit tree showing a document release.
+
+   Example Git tree showing how a document is released and baselined.
+   The release manager creates a ticket branch (``tickets/DM-6``) with two commits from the current release branch (``tickets/RFC-1``).
+   Commit ``D`` updates the document's history table.
+   Commit ``E`` sets the document's version of removes the draft watermark.
```
Comment (timj) [inline docs/change-controlled-docs.rst]: The version number is in the change record so I would normally combine D and E in one commit (we don't have a `setDocVersion` command). In your case you are trying to keep the removal of draft as a separate commit to help with cherry picking.

File: docs/change-controlled-docs.rst [hunk 16]
```diff
@@ -0,0 +1,233 @@
...
+Hotfixing a baselined document
+==============================
+
+It may be necessary to release a minor update to a baselined document (to fix typos, for example).
+This is done by creating a branch based off the prior release branch.
+Hotfixes cannot be made from ``master`` because significant edits may have been merged since the document was baselined.
+
+1. Create a ticket branch from the semantic version tag of the prior release.
+2. Commit fixes to that ticket branch and have the changes peer reviewed.
+3. Create a DocuShare tag (``docushare-vN``) and upload the document to DocuShare.
```
Comment (timj) [inline docs/change-controlled-docs.rst]: standard ordering problem. Maybe we can simplify the repeat of the logic here by saying that this is exactly like a normal CCB submission but the branch point is the prior version tag rather than master?

File: docs/change-controlled-docs.rst [hunk 17]
```diff
@@ -0,0 +1,233 @@
...
+  This means that the DM release manager is responsible for properly addressing conflicts while cherry picking amendments back to ``master``.
+
+.. figure:: ccd-release.svg
+   :name: fig-release
+   :align: center
+   :target: ccd-release.svg
+   :alt: Git commit tree showing a document release.
+
+   Example Git tree showing how a document is released and baselined.
+   The release manager creates a ticket branch (``tickets/DM-6``) with two commits from the current release branch (``tickets/RFC-1``).
```
Comment (timj) [inline docs/change-controlled-docs.rst]: Do we really need to create a JIRA ticket for updating the change record? I'd generally put that directly on the release branch.

File: docs/change-controlled-docs.rst [hunk 18]
```diff
@@ -0,0 +1,233 @@
...
+.. _ccd-docushare-tag:
+
+DocuShare tags
+--------------
+
+DocuShare tags are formatted as ``docushare-vN`` where ``N`` corresponds to a document version number in DocuShare.
+DocuShare version numbers increment by one each time a new version of a document for a given handle is uploaded to DocuShare.
+Note that DocuShare version numbers are distinct from :ref:`semantic version numbers <ccd-semantic-tag>`.
+
+Since DocuShare version numbers are only created once a file is uploaded to DocuShare, we always create DocuShare tags for the next available available DocuShare version, and upload the corresponding PDF to DocuShare.
```
Comment (timj) [inline docs/change-controlled-docs.rst]: "available available"

File: docs/change-controlled-docs.rst [hunk 19]
```diff
@@ -0,0 +1,233 @@
...
+This sequence illustrates:
+
+1. The newest existing version of a document in DocuShare is version 5.
+2. The release manager creates the tag ``docushare-v6`` and uploads the associated PDF built by a continuous integration system to DocuShare.
+3. That upload becomes version 6.
```
Comment (timj) [inline docs/change-controlled-docs.rst]: Wrong order.

File: docs/change-controlled-docs.rst [hunk 20]
```diff
@@ -0,0 +1,233 @@
...
+
+.. _ccd-semantic-tag:
+
+Semantic version tags
+---------------------
+
+Semantic version tags are formatted as ``v<major>.<minor>``.
+The meanings of semantic document versions are described in `LPM-51`_.
+
+Note that only the LSST project librarian can assign a semantic version to a released document.
```
Comment (timj) [inline docs/change-controlled-docs.rst]: Or release manager.

File: docs/change-controlled-docs.rst (hunk 15 above)
Comment (jonathansick) [reply docs/change-controlled-docs.rst]: Ah okay. Sounds goods. I'll change it so the final release happens directly on the release branch.

File: docs/change-controlled-docs.rst (hunk 7 above)
Comment (jonathansick) [reply docs/change-controlled-docs.rst]: Up to you. I can document it as either/or.

File: docs/change-controlled-docs.rst (hunk 4 above)
Comment (jonathansick) [reply docs/change-controlled-docs.rst]: I'm not capitalizing job titles for consistency with Document-13016 § 2.6 Capitalization of Titles.

Who does what is a good question. I'll pull that conversation back to the main comments rather than this line thread.

**Comment by jonathansick** [comment]
@timj So you'd like to have the document uploaded to DocuShare, then get the DocuShare version number from that upload? Okay, I can write it that way. The way it's currently written would allow a DocuShare upload to be triggered whenever CI sees a new `docushare-vN` tag, but I'm sure we could devise a separate process.

The other bit of feedback I'm seeing is about who does what. It sounds like for LDM documents we can have any individual in DM create the RFC. And if we use the lsst.io sites for DM RFCs, all the better. Then the main function of the DM release manager is to create the final DocuShare tag and semantic version tag at the end of the approval process and deal with the DocuShare upload.

for LSE and LPM documents, should we say that the DM release manager (which could be a class of people that includes you, how do we phrase that?) creates the LCR, and therefore also creates the `tickets/LCR-N` release branch.

**Comment by timj** [comment]
> So you'd like to have the document uploaded to DocuShare, then get the DocuShare version number from that upload?

I'm worried about race conditions on the rare occasion where someone uploads a different draft of the LDM after the user has tagged but before they themselves uploaded. With LSE it's even worse because you don't really know what Rob is going to do between you tagging and him uploading your version. We'd have to talk to Rob if we want LSE drafts automatically uploaded.

As for who does what, yes DM CCB is a lot more flexible and anyone can create an RFC for an LDM even if they don't themselves have permission to upload to docushare. In theory LPM-51 would like the draft to turn up on docushare so the people managing LDM releases might have to be involved for that bit. "DM Systems Engineering Team" might be okay for the team that can release final versions of documents.

**Comment by jonathansick** [comment]
@timj I've re-written the procedure and now I'm ready for you to review it again.

File: docs/change-controlled-docs.rst [hunk 21]
```diff
@@ -0,0 +1,293 @@
+.. _writing-ccd:
+
+#######################################
+Writing change-controlled documentation
+#######################################
+
+Change-controlled documents are documents that require approval from a Change Control Board (CCB) before they can be released.
```
Comment (timj) [inline docs/change-controlled-docs.rst]: We probably should have the note at the top declaring that this page is change-controlled.

File: docs/change-controlled-docs.rst [hunk 22]
```diff
@@ -0,0 +1,293 @@
...
+Instead, it adds to the changeset that will be included in the next CCB review (see next).
+
+.. _ccd-release:
+
+Releasing a new version from the master branch
+==============================================
+
+Follow these steps to submit a document to the CCB and release a new baselined version:
+
+1. Check out the head of the ``master`` branch and create a new DocuShare upload from the generated PDF.
```
Comment (timj) [inline docs/change-controlled-docs.rst]: somehow this sentence has to make clear that "creating a new Docushare upload" may well mean "send to project librarian or release manager".

Maybe: "... and ensure the generated PDF is uploaded to DocuShare." ?

File: docs/change-controlled-docs.rst [hunk 23]
```diff
@@ -0,0 +1,293 @@
...
+
+   Repeat this step for each round of CCB feedback.
+
+5. When the CCB approves document you create a release:
+
+   1. Commit to the head of the release branch with the following changes:
+
+      - Update `document's change record <https://lsst-texmf.lsst.io/lsstdoc.html#document-preamble>`_.
+        The Project librarian or DM release manager, through the CCB, determines the document's semantic version.
+      - Remove the ``draft`` option from the document class.
```
Comment (timj) [inline docs/change-controlled-docs.rst]: `lsstdraft` option

File: docs/change-controlled-docs.rst [hunk 24]
```diff
@@ -0,0 +1,293 @@
...
+     You don't need to create intermediate DocuShare versions for the DM CCB.
+
+   Repeat this step for each round of CCB feedback.
+
+5. When the CCB approves document you create a release:
+
+   1. Commit to the head of the release branch with the following changes:
+
+      - Update `document's change record <https://lsst-texmf.lsst.io/lsstdoc.html#document-preamble>`_.
+        The Project librarian or DM release manager, through the CCB, determines the document's semantic version.
```
Comment (timj) [inline docs/change-controlled-docs.rst]: Rob will also check the content of the change record before accepting the LSE/LPM document revision so you may need to iterate this step.

File: docs/change-controlled-docs.rst [hunk 25]
```diff
@@ -0,0 +1,293 @@
...
+            git checkout -b u/<username>/v<major>.<minor>-backport
+
+      2. Cherry-pick commits from the release branch onto the new backport branch.
+         For example:
+
+         .. code-block:: bash
+
+            git cherry-pick <commit-sha>
+         
+         Also ensure that the document's change record is updated, but the ``draft`` option must remain in the document class command.
```
Comment (timj) [inline docs/change-controlled-docs.rst]: it might be easier to achieve this if the draft removal and date setting is on a different commit to the change record modification.

File: docs/change-controlled-docs.rst [hunk 26]
```diff
@@ -0,0 +1,293 @@
...
+            git merge --no-ff u/<username>/v<major>.<minor>-backport
+            git push
+
+.. _ccd-hotfix:
+
+Hotfixing a released document
+=============================
+
+The procedure above (:ref:`ccd-release`) describes how to make a new version of a document from the ``master`` branch.
+Sometimes it is necessary to hotfix a released document to fix a typo or make a similar minor change.
```
Comment (timj) [inline docs/change-controlled-docs.rst]: > ...whilst other edits of this document are ongoing. ?

I think that if you are making a change to a document that has no other changes being made to it you can use the normal process.

File: docs/change-controlled-docs.rst [hunk 27]
```diff
@@ -0,0 +1,293 @@
...
+
+   - For project documents (such as LPM and LSE), email the PDF to the LSST librarian.
+
+   - For DM documents (such as LDM and DMTR), email the PDF to a DM documentalist.
+
+2. Wait for the documentalist or librarian to upload the document and verify that it appears on the Version page of the document on DocuShare.
+   You can find the document version page with the short link ``https://ls.st/<handle>*``.
+   For example, `https://ls.st/ldm-151* <https://ls.st/ldm-151*>`_.
+
+3. Tag the commit that produced the DocuShare upload.
```
Comment (timj) [inline docs/change-controlled-docs.rst]: This is duplicating information from earlier in the document. Do we need it again?

Comment (jonathansick) [reply docs/change-controlled-docs.rst]: Could you point to where this is duplicating information? I've arranged it so that whenever a DocuShare upload is called for, this section is referenced. I don't think I describe how to make a DocuShare tag anywhere else.

You might be thinking of the semantic tag, but that's distinct and occurs in only one place.

Let me know.

Comment (timj) [reply docs/change-controlled-docs.rst]: Oops. Yes, I didn't spot the difference as the text was almost identical and I felt like I had a glitch in the Matrix.
